aesara/tensor/math_opt.py

```diff
@@ -2283,37 +2284,32 @@ def local_log1p(fgraph, node):
 
 
 # TODO: in canonicalize, change log10 and log2 -> log
-@register_stabilize
-@register_specialize
 @local_optimizer([log])
-def local_log_add(fgraph, node):
-    # log(exp(x)+exp(y))
-    #
-    # Suppose x >= y
-    # log(exp(x) + exp(y))
-    # log(exp(x) * (1 + exp(y)/exp(x)))
-    # x + log(1 + exp(y)/exp(x))
-    # x + log1p(exp(y)/exp(x))
-    # x + log1p(exp(y-x))
+def local_log_add_exp(fgraph, node):
+    # log(exp(x)+exp(y)+exp(z)) -> max + log(x-max, y-max, z-max)
+
     if node.op == log:
         z = node.inputs[0]
         if z.owner and z.owner.op == add:
             zi = z.owner.inputs
-            if len(zi) != 2:
-                # -- upgrading Maximum to handle multiple inputs wasn't trivial
-                #    TODO
-                # raise NotImplementedError()
-                return
             pre_exp = [x.owner.inputs[0] for x in zi if x.owner and x.owner.op == exp]
+            # all arguments to add are exp(<something>)
             if len(pre_exp) == len(zi):
-                # all arguments to add are exp(<something>)
-                max_pre = maximum(*pre_exp)
-
-                ret = max_pre + log1p(exp(add(*[p - max_pre for p in pre_exp])))
-                ret.tag.values_eq_approx = values_eq_approx_remove_inf
+                max_pre = reduce(maximum, pre_exp)
+                # If max == -inf, offset would lead to nan output
+                max_pre = switch(isinf(max_pre), 0, max_pre)
+                ret = max_pre + log(add(*[exp(p - max_pre) for p in pre_exp]))
```

Comment:
The result was previously returned as a `logp1`.  That change is what prevents the fixed-point (aka "equilibrium") optimizers (e.g. the optimizer behind "stabilize") from converging, since it produces yet another `log(exp(...) + exp(...))` form into the graph, and that matches the implicit "pattern" of `local_log_add_exp`, etc., etc.

We can rewrite this so that it doesn't return the same form it replaces, or we can attempt to determine whether or not this `node` is already in the desired form (i.e. add a guard).  The former is generally easier and more consistent.

Reply:
Thanks for the suggestion, I think there is an inherent limitation to the `log1p` trick, that does not allow for the switch safeguard nor even for more than two inputs, but I have to double check.

Reply:
See [here](https://github.com/pymc-devs/aesara/pull/465#discussion_r646232449).

Reply:
For instance if you have three inputs of `0` being exp added the log1p trick fails already

Reply:
Maybe!